svm/src/transaction_processor.rs

```diff
@@ -1292,6 +1299,165 @@ mod tests {
         );
     }
 
+    #[test_case(0; "Throw in program 0")]
+    #[test_case(1; "Throw in program 1")]
+    #[test_case(2; "Throw in program 2")]
+    #[test_case(3; "Throw in program 3")]
+    #[test_case(4; "Throw in program 4")]
+    #[test_case(5; "Throw in program 5")]
+    #[test_case(6; "Throw in program 6")]
+    fn test_instruction_error_carries_responsible_program_account_index(
+        index_of_program_that_should_throw_exception: u8,
+    ) {
+        lazy_static! {
+            static ref PROGRAM_ADDRESSES: [Pubkey; 7] =
+                std::array::from_fn(|_| Pubkey::new_unique());
+        }
+        // This mock program takes in a list of programs and two bytes of data:
+        //   (1) the index of the program that should throw an error
+        //   (2) the index of the program being called
+        //
+        // The programs get executed - two at each CPI call depth - like this:
+        //
+        // Program 0
+        // -- Program 1
+        // -- Program 2
+        // ---- Program 3
+        // ---- Program 4
+        // ------ Program 5
+        // ------ Program 6
```

Comment:
This is the meat of the test. Essentially I created a program that calls itself over and over as diagrammed in the comment. You can pass it the index at which you want it to throw an error. The test then makes sure that the `TransactionError::InstructionError` indicates that the error was thrown from the program whose account index is that one exactly.